### PR TITLE
Avoid always locking in BearerTokenAuthenticationPolicy

### PR DESCRIPTION
**Pull request by johnthcall**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

Moves token cache state to struct so that state can be checked outside the lock and return if there is already a valid token (99% case for high throughput clients)

### CODE REVIEW

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (AlexanderSher) [inline sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:290]: Reading a three field struct isn't atomic:
```cs
internal class Program
{
    private readonly object _syncObj = new object();
    private ThreeFields _str;

    static void Main(string[] args)
    {
        new Program().Start();
    }

    private void Start()
    {
        for (var i = 0; i < Environment.ProcessorCount - 1; i++)
        {
            ThreadPool.QueueUserWorkItem(Read, null);
        }
        ThreadPool.QueueUserWorkItem(Write, null);
        Console.ReadLine();
    }

    private void Read(object state)
    {
        while (true)
        {
            var str = _str;
            Validate(str);
        }
    }

    private void Write(object state)
    {
        while (true)
        {
            lock (_syncObj)
            {
                var count = _str.FieldOne + 1;
                _str = new ThreeFields(count, count, count);
            }
        }
    }

    private static void Validate(ThreeFields str)
    {
        if (str.FieldOne != str.FieldTwo || str.FieldOne != str.FieldThree || str.FieldTwo != str.FieldThree)
        {
            Console.WriteLine($"Field one: {str.FieldOne}, Field two: {str.FieldTwo}, Field three: {str.FieldThree}");
        }
    }

    private readonly struct ThreeFields
    {
        public ThreeFields(int fieldOne, int fieldTwo, int fieldThree)
        {
            FieldOne = fieldOne;
            FieldTwo = fieldTwo;
            FieldThree = fieldThree;
        }

        public int FieldOne { get; }
        public int FieldTwo { get; }
        public int FieldThree { get; }
    }
}
```

Comment (johnthcall) [reply sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:290]: Thanks for the example will change to a class to get atomicity

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (AlexanderSher) [inline sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:293]: You can add `Deconstruct` to `TokenRequestState` to use individual variables for each state item

`RequestRequiresNewToken` can be checked last:
```cs
infoTcs != null && backgroundUpdateTcs == null && infoTcs.Task.Status == TaskStatus.RanToCompletion && !RequestRequiresNewToken(context, currentContext)
```

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (AlexanderSher) [inline sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:314]: New instance of `TokenRequestState` is required only if `BackgroundUpdateTcs != null`.

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (johnthcall) [reply sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs]: I also used Deconstruct inside of the lock, let me know if you'd prefer to keep it as it was.

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (pakrym) [inline sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs:330]: Should we move these conditions to TokenRequestState and make them properties. Might make this method logic a bit simpler and remove the duplication with the fast path.

Comment (johnthcall) [reply sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs]: Made them methods as they either take in the incoming TokenRequestContext or DateTimeOffset. We could make the ones taking DateTimeOffset into properties and each of the 3 checks would fetch their own DateTimeOffset.UtcNow. Let me know what you'd like me to do.

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs
Comment (johnthcall) [reply sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs]: After moving functionality to Properties/Methods I have reverted the Deconstruct, let me know if you think it still makes sense to have it.

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
@AlexanderSher do you want to take the last look?

**Comment by johnthcall** [comment]
@pakrym, do we need a second review for this change to go in?

**Review by AlexanderSher — APPROVED** [review]

**Pull request merged**
